loadtest/agentconn/run.go

```diff
@@ -0,0 +1,321 @@
+		}
+
+		select {
+		case <-ctx.Done():
+			return xerrors.Errorf("verify connection: %w", ctx.Err())
+		case <-time.After(verifyConnectionDelay):
+		}
+	}
+
+	_, _ = fmt.Fprint(logs, "\nConnection verified.\n\n")
+
+	// Make initial connections sequentially to ensure the services are
+	// reachable before we start spawning a bunch of goroutines and tickers.
+	if len(r.cfg.Connections) > 0 {
+		_, _ = fmt.Fprintln(logs, "Performing initial service connections...")
+	}
+	for i, connSpec := range r.cfg.Connections {
+		_, _ = fmt.Fprintf(logs, "\t%d. %s\n", i, connSpec.URL)
+
+		timeout := defaultRequestTimeout
+		if connSpec.Timeout > 0 {
+			timeout = time.Duration(connSpec.Timeout)
+		}
+		ctx, cancel := context.WithTimeout(ctx, timeout)
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, connSpec.URL, nil)
+		if err != nil {
+			cancel()
+			return xerrors.Errorf("create request: %w", err)
+		}
+
+		res, err := client.Do(req)
+		cancel()
+		if err != nil {
+			_, _ = fmt.Fprintf(logs, "\t\tFailed: %+v\n", err)
+			return xerrors.Errorf("make initial connection to conn spec %d %q: %w", i, connSpec.URL, err)
+		}
+		_ = res.Body.Close()
+
+		_, _ = fmt.Fprintln(logs, "\t\tOK")
+	}
+
+	if r.cfg.HoldDuration > 0 {
+		eg, egCtx := errgroup.WithContext(ctx)
+
+		if len(r.cfg.Connections) > 0 {
+			_, _ = fmt.Fprintln(logs, "\nStarting connection loops...")
+		}
+		for i, connSpec := range r.cfg.Connections {
+			i, connSpec := i, connSpec
+			if connSpec.Interval <= 0 {
+				continue
+			}
+
+			eg.Go(func() error {
+				t := time.NewTicker(time.Duration(connSpec.Interval))
+				defer t.Stop()
+
+				timeout := defaultRequestTimeout
+				if connSpec.Timeout > 0 {
+					timeout = time.Duration(connSpec.Timeout)
+				}
+
+				for {
+					select {
+					case <-egCtx.Done():
+						return egCtx.Err()
+					case <-t.C:
+						ctx, cancel := context.WithTimeout(ctx, timeout)
+						req, err := http.NewRequestWithContext(ctx, http.MethodGet, connSpec.URL, nil)
+						if err != nil {
+							cancel()
+							return xerrors.Errorf("create request: %w", err)
+						}
+
+						res, err := client.Do(req)
+						cancel()
+						if err != nil {
+							_, _ = fmt.Fprintf(logs, "\tERR: %s (%d): %+v\n", connSpec.URL, i, err)
+							return xerrors.Errorf("make connection to conn spec %d %q: %w", i, connSpec.URL, err)
+						}
+						res.Body.Close()
+
+						_, _ = fmt.Fprintf(logs, "\tOK: %s (%d)\n", connSpec.URL, i)
+						t.Reset(time.Duration(connSpec.Interval))
+					}
+				}
+			})
+		}
+
+		// Wait for the hold duration to end. We use a fake error to signal that
+		// the hold duration has ended.
+		_, _ = fmt.Fprintf(logs, "\nWaiting for %s...\n", time.Duration(r.cfg.HoldDuration))
+		eg.Go(func() error {
+			t := time.NewTicker(time.Duration(r.cfg.HoldDuration))
+			defer t.Stop()
+
+			select {
+			case <-egCtx.Done():
+				return egCtx.Err()
+			case <-t.C:
+				return holdDurationEndedError{}
```

Comment:
Using the errgroup context is pretty slick for cancellation. I think adding a comment for why we return this error would be helpful, took me a few mins to wrap my head around it.